### ISSUE TITLE
Terraform init does not cache the location of a specific module version from the Terraform Registry

### ISSUE THREAD

**Issue by pkolyvas** [hashicorp/terraform#23544]
### Terraform Version

```
Terraform v0.12.17
```

### Terraform Configuration Files
```hcl
module "security-group-reg-1" {
  source  = "terraform-aws-modules/security-group/aws"
  version = "3.2.0"
  name = "sg-1"
  vpc_id = aws_vpc.state-vpc.id
}

module "security-group-reg-2" {
  source  = "terraform-aws-modules/security-group/aws"
  version = "3.2.0"
  name = "sg-2"
  vpc_id = aws_vpc.state-vpc.id
}

module "security-group-reg-3" {
  source  = "terraform-aws-modules/security-group/aws"
  version = "3.2.0"
  name = "sg-3"
  vpc_id = aws_vpc.state-vpc.id
}

module "security-group-git-1" {
  source  = "github.com/terraform-aws-modules/terraform-aws-security-group.git"
  name = "sg-4"
  vpc_id = aws_vpc.state-vpc.id
}

module "security-group-git-2" {
  source  = "github.com/terraform-aws-modules/terraform-aws-security-group.git"
  name = "sg-5"
  vpc_id = aws_vpc.state-vpc.id
}

module "security-group-git-3" {
  source  = "github.com/terraform-aws-modules/terraform-aws-security-group.git"
  name = "sg-6"
  vpc_id = aws_vpc.state-vpc.id
}
```

### Debug Output

https://gist.github.com/pkolyvas/6e80adbabe0569773880d1357194e8c4

### Expected Behavior
Terraform should have fetched a single copy of the module from the Terraform Module Registry and a single copy of the module from GitHub.

### Actual Behavior
Terraform fetched three copies of the module from the Terraform Module Registry and a single copy of the module from GitHub.

### Steps to Reproduce
1. Use multiple instances of the same module from the Terraform Module Registry in a given `.tf` config.
2. `terraform init`

### Additional context

On very large configurations/plans, repeatedly downloading the same, large, modules can dramatically affect performance.

**Comment by apparentlymart** [comment]
Unfortunately I've been unable to reproduce the bug described here. When I ran the given configuration with a local copy of Terraform (built from latest `master`) I saw log lines indicating that Terraform is correctly re-using the existing directory for the second registry module call once it's recognized that they are both referring to the same source:

First call:
```
2019/12/03 14:13:41 [DEBUG] looking up module location from "https://registry.terraform.io/v1/modules/terraform-aws-modules/security-group/aws/3.2.0/download"
2019/12/03 14:13:41 [TRACE] HTTP client GET request to https://registry.terraform.io/v1/modules/terraform-aws-modules/security-group/aws/3.2.0/download
Downloading terraform-aws-modules/security-group/aws 3.2.0 for security-group-reg-1...
2019/12/03 14:13:41 [TRACE] ModuleInstaller: security-group-reg-1 terraform-aws-modules/security-group/aws 3.2.0 is available at "https://api.github.com/repos/terraform-aws-modules/terraform-aws-security-group/tarball/v3.2.0//*?archive=tar.gz"
2019/12/03 14:13:41 [DEBUG] will download "https://api.github.com/repos/terraform-aws-modules/terraform-aws-security-group/tarball/v3.2.0?archive=tar.gz" to .terraform/modules/security-group-reg-1
2019/12/03 14:13:41 [TRACE] fetching "https://api.github.com/repos/terraform-aws-modules/terraform-aws-security-group/tarball/v3.2.0?archive=tar.gz" to ".terraform/modules/security-group-reg-1"
2019/12/03 14:13:42 [TRACE] expanded "*" to ".terraform/modules/security-group-reg-1/terraform-aws-modules-terraform-aws-security-group-57856b3"
2019/12/03 14:13:42 [TRACE] ModuleInstaller: security-group-reg-1 "https://api.github.com/repos/terraform-aws-modules/terraform-aws-security-group/tarball/v3.2.0//*?archive=tar.gz" was downloaded to .terraform/modules/security-group-reg-1/terraform-aws-modules-terraform-aws-security-group-57856b3
2019/12/03 14:13:42 [TRACE] ModuleInstaller: security-group-reg-1 should now be at .terraform/modules/security-group-reg-1/terraform-aws-modules-terraform-aws-security-group-57856b3
2019/12/03 14:13:42 [DEBUG] Module installer: security-group-reg-1 installed at .terraform/modules/security-group-reg-1/terraform-aws-modules-terraform-aws-security-group-57856b3
```
Second call:
```
2019/12/03 14:13:42 [DEBUG] looking up module location from "https://registry.terraform.io/v1/modules/terraform-aws-modules/security-group/aws/3.2.0/download"
2019/12/03 14:13:42 [TRACE] HTTP client GET request to https://registry.terraform.io/v1/modules/terraform-aws-modules/security-group/aws/3.2.0/download
Downloading terraform-aws-modules/security-group/aws 3.2.0 for security-group-reg-2...
2019/12/03 14:13:42 [TRACE] ModuleInstaller: security-group-reg-2 terraform-aws-modules/security-group/aws 3.2.0 is available at "https://api.github.com/repos/terraform-aws-modules/terraform-aws-security-group/tarball/v3.2.0//*?archive=tar.gz"
2019/12/03 14:13:42 [DEBUG] will download "https://api.github.com/repos/terraform-aws-modules/terraform-aws-security-group/tarball/v3.2.0?archive=tar.gz" to .terraform/modules/security-group-reg-2
2019/12/03 14:13:42 [TRACE] copying previous install .terraform/modules/security-group-reg-1 to .terraform/modules/security-group-reg-2
2019/12/03 14:13:42 [TRACE] expanded "*" to ".terraform/modules/security-group-reg-2/terraform-aws-modules-terraform-aws-security-group-57856b3"
2019/12/03 14:13:42 [TRACE] ModuleInstaller: security-group-reg-2 "https://api.github.com/repos/terraform-aws-modules/terraform-aws-security-group/tarball/v3.2.0//*?archive=tar.gz" was downloaded to .terraform/modules/security-group-reg-2/terraform-aws-modules-terraform-aws-security-group-57856b3
2019/12/03 14:13:42 [TRACE] ModuleInstaller: security-group-reg-2 should now be at .terraform/modules/security-group-reg-2/terraform-aws-modules-terraform-aws-security-group-57856b3
2019/12/03 14:13:42 [DEBUG] Module installer: security-group-reg-2 installed at .terraform/modules/security-group-reg-2/terraform-aws-modules-terraform-aws-security-group-57856b3
```

The important difference between these two is the log line after "will download..." in each of them:

```
[TRACE] fetching "https://api.github.com/repos/terraform-aws-modules/terraform-aws-security-group/tarball/v3.2.0?archive=tar.gz" to ".terraform/modules/security-group-reg-1"
```

```
[TRACE] copying previous install .terraform/modules/security-group-reg-1 to .terraform/modules/security-group-reg-2
```

It seems that [the attached trace log](https://gist.github.com/pkolyvas/6e80adbabe0569773880d1357194e8c4) is missing some log lines I'd expect to see. For example, the string "will download" appears only twice, but I'd expect to see it six times because Terraform emits that log line each time it's about to fetch a module, before it does any other work to normalize the source address and check the cache.

@pkolyvas, did you edit the log you shared at all, or was the exact output from your Terraform run? If the latter, could you share with me the exact command line you used to produce it, so that I can try to reproduce? Thanks!

**Comment by pkolyvas** [comment]
@apparentlymart I ran Terraform at `TF_LOG=DEBUG` shall I run it with `TRACE`?

**Comment by apparentlymart** [comment]
Confirmed with @pkolyvas in an offline discussion that his log was incomplete due to being a `DEBUG` log, and was strangely subsetted due to #6468.

With that said, it seems like Terraform is behaving as designed here: it is correctly reusing a previously-downloaded directory once it has gathered enough information to prove that the source location would be the same.

I think what we have here is instead a feature request for some additional caching levels further up. For example:
* Once we've fetched the list of available versions for a particular registry module, we could potentially remember that answer and re-use it for subsequent calls that have the same normalized registry `source` address.
* Once we've asked the registry for the download location for a particular version of a particular registry module, we could potentially remember that answer and re-use it for subsequent calls that are for the same version of the same module.

If we were to do both of the above then Terraform could avoid get away with making only two registry API calls with the configuration given in the example: one call to fetch all of the versions for `registry.terraform.io/terraform-aws-modules/security-group/aws`, and then one call to find the download location for version `3.2.0`.

We could potentially optimize further by recognizing that this version constraint is selecting a single exact version, skip over the step of lookup up which versions of the module are available, and just try asking directly for the download location for version `3.2.0`.

The current registry interaction codepaths are not equipped to retain enough state to cache information in that way, so it would take some refactoring to do so. I believe we could do it by adding an internal cache (or two) to `initwd.ModuleInstaller` which it would consult before attempting to install a registry module, and then the result of those cached lookups could then feed into our existing cache of which sources we've already downloaded with `go-getter`.
